parquet/src/main/java/org/apache/iceberg/parquet/ParquetFileMerger.java

```diff
@@ -0,0 +1,643 @@
+   *
+   * <p>Row lineage consists of two columns:
+   *
+   * <ul>
+   *   <li>_row_id: unique identifier for each row
+   *   <li>_last_updated_sequence_number: data sequence number when row was last updated
+   * </ul>
+   *
+   * @param baseSchema Original Parquet schema
+   * @return Extended schema with row lineage columns added
+   */
+  private static MessageType addRowLineageColumns(MessageType baseSchema) {
+    // Create _row_id column: required int64 with Iceberg field ID for proper column mapping
+    PrimitiveType rowIdType =
+        Types.required(PrimitiveType.PrimitiveTypeName.INT64)
+            .id(MetadataColumns.ROW_ID.fieldId())
+            .named(MetadataColumns.ROW_ID.name());
+
+    // Create _last_updated_sequence_number column: required int64 with Iceberg field ID
+    PrimitiveType seqNumType =
+        Types.required(PrimitiveType.PrimitiveTypeName.INT64)
+            .id(MetadataColumns.LAST_UPDATED_SEQUENCE_NUMBER.fieldId())
+            .named(MetadataColumns.LAST_UPDATED_SEQUENCE_NUMBER.name());
+
+    // Add to existing fields
+    List<Type> fields = Lists.newArrayList(baseSchema.getFields());
+    fields.add(rowIdType);
+    fields.add(seqNumType);
+
+    return new MessageType(baseSchema.getName(), fields);
+  }
+
+  /**
+   * Writes a _row_id column chunk with sequential row IDs.
+   *
+   * <p>Uses DELTA_BINARY_PACKED encoding for sequential data, where deltas are all 1.
+   *
+   * @param writer ParquetFileWriter to write to
+   * @param rowIdDescriptor Column descriptor for _row_id
+   * @param startRowId Starting row ID for this row group
+   * @param rowCount Number of rows in this row group
+   * @param codec Compression codec to use (should match the file's codec)
+   * @throws IOException if writing fails
+   */
+  private static void writeRowIdColumnChunk(
+      ParquetFileWriter writer,
+      ColumnDescriptor rowIdDescriptor,
+      long startRowId,
+      long rowCount,
+      CompressionCodecName codec)
+      throws IOException {
+    // Write sequential values: startRowId, startRowId+1, startRowId+2, ...
+    writeLongColumnChunk(
+        writer,
+        rowIdDescriptor,
+        rowCount,
+        codec,
+        startRowId,
+        startRowId + rowCount - 1,
+        i -> startRowId + i);
+  }
+
+  /**
+   * Writes a _last_updated_sequence_number column chunk with constant sequence number.
+   *
+   * <p>Uses DELTA_BINARY_PACKED encoding for constant data, where deltas are all 0.
+   *
+   * @param writer ParquetFileWriter to write to
+   * @param seqNumDescriptor Column descriptor for _last_updated_sequence_number
+   * @param dataSequenceNumber Data sequence number for all rows in this file
+   * @param rowCount Number of rows in this row group
+   * @param codec Compression codec to use (should match the file's codec)
+   * @throws IOException if writing fails
+   */
+  private static void writeSequenceNumberColumnChunk(
```

Comment:
We don't need a this method. Inline the call

Reply:
Sure